### ISSUE TITLE
使用paddleinference部署，推理时间反而比动态图推理时间长

### ISSUE THREAD

**Issue by CCZGG** [PaddlePaddle/PaddleNLP#983]
欢迎您反馈PaddleNLP使用问题，非常感谢您对PaddleNLP的贡献！
在留下您的问题时，辛苦您同步提供如下信息：
- 版本、环境信息
1）PaddleNLP和PaddlePaddle版本：请提供您的PaddleNLP和PaddlePaddle版本号，例如PaddleNLP 2.0.4，PaddlePaddle2.1.1
2）系统环境：请您描述系统类型，例如Linux/Windows/MacOS/，python版本
- 复现信息：如为报错，请给出复现环境、复现步骤

<img width="906" alt="cab6d47c79b8090c81c66a7b6aa0e52" src="https://user-images.githubusercontent.com/44726956/132447587-37082b4f-0c9a-4d33-a685-bab5a39580ca.png">

ubuntu18.04
paddle版本是：python -m pip install paddlepaddle-gpu==0.0.0.post112 -f https://www.paddlepaddle.org.cn/whl/linux/gpu/develop.html

模型：bigru-crf

使用的paddleinference的python接口

疑问：为什么推理时间比动态图更长？怎么解决？

**Comment by wawltor** [comment]
这个问题的原因是在动转静的过程中会将很多动态图的for if这样的控制流转成op的方式来执行，在CRF里面又大量存在这样的控制流操作，导致动转静之后速度会有问题，这个问题我们在推动PaddlePaddle框架来解决。

**Comment by CCZGG** [comment]
感谢解疑

**Comment by CCZGG** [comment]
> 这个问题的原因是在动转静的过程中会将很多动态图的for if这样的控制流转成op的方式来执行，在CRF里面又大量存在这样的控制流操作，导致动转静之后速度会有问题，这个问题我们在推动PaddlePaddle框架来解决。

还有个问题是，转化为静态图后，原本在动态图抽取正确的数据，在静态图抽取结果会出错。

**Comment by CCZGG** [comment]
> 这个问题的原因是在动转静的过程中会将很多动态图的for if这样的控制流转成op的方式来执行，在CRF里面又大量存在这样的控制流操作，导致动转静之后速度会有问题，这个问题我们在推动PaddlePaddle框架来解决。

paddleinference代码如下：

`import os
import paddle
from functools import partial
from paddlenlp.datasets import MapDataset
from paddlenlp.data import Stack, Tuple, Pad
from data import load_dict
import argparse
import paddle.inference as paddle_infer
import numpy as np

parser = argparse.ArgumentParser(__doc__)
parser.add_argument("--model_dir", type=str, default='./output/',
                    help="The path of model parameter to be loaded.")
parser.add_argument("--data_dir", type=str, default='./data/',
                    help="The path of model parameter to be loaded.")
args = parser.parse_args()


def load_text(text_list):
    def read(text):
        for line in text_list:
            words = line.strip('\n')
            words = [i for i in words]
            # labels = ["O" for i in words]
            yield words
    return MapDataset(list(read(text_list)))


def convert_tokens_to_ids(tokens, vocab, oov_token=None):
    token_ids = []
    oov_id = vocab.get(oov_token) if oov_token else None
    for token in tokens:
        token_id = vocab.get(token, oov_id)
        token_ids.append(token_id)
    return token_ids


def convert_to_features(example, word_vocab):
    tokens = example
    token_ids = convert_tokens_to_ids(tokens, word_vocab, 'OOV')
    return token_ids, len(token_ids)


def parse_decodes(sentences, predictions, lengths, label_vocab):
    """Parse the padding result
    Args:
        sentences (list): the tagging sentences.
        predictions (list): the prediction tags.
        lengths (list): the valid length of each sentence.
        label_vocab (dict): the label vocab.
    Returns:
        outputs (list): the formatted output.
    """
    predictions = [x for batch in predictions for x in batch]
    lengths = [x for batch in lengths for x in batch]
    id_label = dict(zip(label_vocab.values(), label_vocab.keys()))

    outputs = []
    for idx, end in enumerate(lengths):
        sent = sentences[idx][:end]
        tags = [id_label[x] for x in predictions[idx][:end]]
        print(tags)
        sent_out = []
        tags_out = []
        words = ""
        for s, t in zip(sent, tags):
            if t.endswith('-B') or t == 'O':
                if len(words):
                    sent_out.append(words)
                tags_out.append(t.split('-')[0])
                words = s
            else:
                words += s
        if len(sent_out) < len(tags_out):
            sent_out.append(words)
        # outputs.append(''.join(
        #     [str((s, t)) for s, t in zip(sent_out, tags_out)]))
        outputs.append([(s, t) for s, t in zip(sent_out, tags_out)])
    return outputs


class InfoExtract(object):
    def __init__(self):
        paddle.set_device('gpu')
        self.word_vocab = load_dict('./data/word.dic')
        self.model_file = os.path.join(args.model_dir, "inference.pdmodel")
        self.params_file = os.path.join(args.model_dir, "inference.pdiparams")
        self.label_vocab = load_dict(os.path.join(args.data_dir, 'tag.dic'))
        # 创建 config，并设置预测模型路径
        config = paddle_infer.Config(self.model_file, self.params_file)
        self.predictor = paddle_infer.create_predictor(config)
        self.input_handles = [
            self.predictor.get_input_handle(name)
            for name in self.predictor.get_input_names()
        ]
        self.output_handle = self.predictor.get_output_handle(
            self.predictor.get_output_names()[0])
        print("************** bigru-crf predictor loaded success ******************")

    def data_process(self, text_list):
        ds = load_text(text_list)
        trans_func = partial(
            convert_to_features, word_vocab=self.word_vocab)
        ds.map(trans_func)
        batchify_fn = lambda samples, fn=Tuple(
            Pad(axis=0, pad_val=self.word_vocab.get('OOV', 0), dtype='int32'),  # token_ids
            Stack(dtype='int64'),  # seq_len
            # Pad(axis=0, pad_val=label_vocab.get('O', 0), dtype='int64')  # label_ids
        ): fn(samples)
        input_id, length = batchify_fn(ds)
        return input_id, length, ds

    def predict(self, text):
        all_preds = []
        all_lens = []
        if len(text) > 1:
            text_list = [text.strip()]
            input_id, length, dataset = self.data_process(text_list)
            self.input_handles[0].copy_from_cpu(input_id)
            self.input_handles[1].copy_from_cpu(length)
            self.predictor.run()
            logits = self.output_handle.copy_to_cpu()
            all_preds.append(logits)
            all_lens.append(length)
            sentences = [example for example in dataset.data]
            results = parse_decodes(sentences, all_preds, all_lens, self.label_vocab)
            results = [i for res in results for i in res if i[1] in ['CZ', 'DC', 'GD', 'XHJ', 'BHDC', 'BHGD']]
            return results
        else:
            return '文本太短'


if __name__ == '__main__':
    info_extract = InfoExtract()
    while True:
        text = input('请输入文本')
        res = info_extract.predict(text)
        print(text)
        print(res)`

**Comment by CCZGG** [comment]
> 这个问题的原因是在动转静的过程中会将很多动态图的for if这样的控制流转成op的方式来执行，在CRF里面又大量存在这样的控制流操作，导致动转静之后速度会有问题，这个问题我们在推动PaddlePaddle框架来解决。

代码如下：

`
import os
import paddle
from functools import partial
from paddlenlp.datasets import MapDataset
from paddlenlp.data import Stack, Tuple, Pad
from data import load_dict
import argparse
import paddle.inference as paddle_infer
import numpy as np

parser = argparse.ArgumentParser(__doc__)
parser.add_argument("--model_dir", type=str, default='./output/',
                    help="The path of model parameter to be loaded.")
parser.add_argument("--data_dir", type=str, default='./data/',
                    help="The path of model parameter to be loaded.")
args = parser.parse_args()


def load_text(text_list):
    def read(text):
        for line in text_list:
            words = line.strip('\n')
            words = [i for i in words]
            # labels = ["O" for i in words]
            yield words
    return MapDataset(list(read(text_list)))


def convert_tokens_to_ids(tokens, vocab, oov_token=None):
    token_ids = []
    oov_id = vocab.get(oov_token) if oov_token else None
    for token in tokens:
        token_id = vocab.get(token, oov_id)
        token_ids.append(token_id)
    return token_ids


def convert_to_features(example, word_vocab):
    tokens = example
    token_ids = convert_tokens_to_ids(tokens, word_vocab, 'OOV')
    return token_ids, len(token_ids)


def parse_decodes(sentences, predictions, lengths, label_vocab):
    """Parse the padding result
    Args:
        sentences (list): the tagging sentences.
        predictions (list): the prediction tags.
        lengths (list): the valid length of each sentence.
        label_vocab (dict): the label vocab.
    Returns:
        outputs (list): the formatted output.
    """
    predictions = [x for batch in predictions for x in batch]
    lengths = [x for batch in lengths for x in batch]
    id_label = dict(zip(label_vocab.values(), label_vocab.keys()))

    outputs = []
    for idx, end in enumerate(lengths):
        sent = sentences[idx][:end]
        tags = [id_label[x] for x in predictions[idx][:end]]
        print(tags)
        sent_out = []
        tags_out = []
        words = ""
        for s, t in zip(sent, tags):
            if t.endswith('-B') or t == 'O':
                if len(words):
                    sent_out.append(words)
                tags_out.append(t.split('-')[0])
                words = s
            else:
                words += s
        if len(sent_out) < len(tags_out):
            sent_out.append(words)
        # outputs.append(''.join(
        #     [str((s, t)) for s, t in zip(sent_out, tags_out)]))
        outputs.append([(s, t) for s, t in zip(sent_out, tags_out)])
    return outputs


class InfoExtract(object):
    def __init__(self):
        paddle.set_device('gpu')
        self.word_vocab = load_dict('./data/word.dic')
        self.model_file = os.path.join(args.model_dir, "inference.pdmodel")
        self.params_file = os.path.join(args.model_dir, "inference.pdiparams")
        self.label_vocab = load_dict(os.path.join(args.data_dir, 'tag.dic'))
        # 创建 config，并设置预测模型路径
        config = paddle_infer.Config(self.model_file, self.params_file)
        self.predictor = paddle_infer.create_predictor(config)
        self.input_handles = [
            self.predictor.get_input_handle(name)
            for name in self.predictor.get_input_names()
        ]
        self.output_handle = self.predictor.get_output_handle(
            self.predictor.get_output_names()[0])
        print("************** bigru-crf predictor loaded success ******************")

    def data_process(self, text_list):
        ds = load_text(text_list)
        trans_func = partial(
            convert_to_features, word_vocab=self.word_vocab)
        ds.map(trans_func)
        batchify_fn = lambda samples, fn=Tuple(
            Pad(axis=0, pad_val=self.word_vocab.get('OOV', 0), dtype='int32'),  # token_ids
            Stack(dtype='int64'),  # seq_len
            # Pad(axis=0, pad_val=label_vocab.get('O', 0), dtype='int64')  # label_ids
        ): fn(samples)
        input_id, length = batchify_fn(ds)
        return input_id, length, ds

    def predict(self, text):
        all_preds = []
        all_lens = []
        if len(text) > 1:
            text_list = [text.strip()]
            input_id, length, dataset = self.data_process(text_list)
            self.input_handles[0].copy_from_cpu(input_id)
            self.input_handles[1].copy_from_cpu(length)
            self.predictor.run()
            logits = self.output_handle.copy_to_cpu()
            all_preds.append(logits)
            all_lens.append(length)
            sentences = [example for example in dataset.data]
            results = parse_decodes(sentences, all_preds, all_lens, self.label_vocab)
            results = [i for res in results for i in res if i[1] in ['CZ', 'DC', 'GD', 'XHJ', 'BHDC', 'BHGD']]
            return results
        else:
            return '文本太短'


if __name__ == '__main__':
    info_extract = InfoExtract()
    while True:
        text = input('请输入文本')
        res = info_extract.predict(text)
        print(text)
        print(res)



`

**Comment by wawltor** [comment]
> > 这个问题的原因是在动转静的过程中会将很多动态图的for if这样的控制流转成op的方式来执行，在CRF里面又大量存在这样的控制流操作，导致动转静之后速度会有问题，这个问题我们在推动PaddlePaddle框架来解决。
> 
> 还有个问题是，转化为静态图后，原本在动态图抽取正确的数据，在静态图抽取结果会出错。

想问一下 是不是静态图的预测的结果比动态图少一个token？

**Comment by CCZGG** [comment]
> > > 这个问题的原因是在动转静的过程中会将很多动态图的for if这样的控制流转成op的方式来执行，在CRF里面又大量存在这样的控制流操作，导致动转静之后速度会有问题，这个问题我们在推动PaddlePaddle框架来解决。
> > 
> > 
> > 还有个问题是，转化为静态图后，原本在动态图抽取正确的数据，在静态图抽取结果会出错。
> 
> 想问一下 是不是静态图的预测的结果比动态图少一个token？

少一个token的现象遇到了，更新了版本后解决了。现在不少了。出错是个别抽取会出错
